### DISCUSSION TITLE
bbr.connectors

### DISCUSSION THREAD

**Discussion by barrettk** [metrumresearchgroup/rbabylon#598] [Ideas]
### Proposal for a new R package

We had a request to add a `bbr` feature for converting already run `PsN` models. The user was interested in moving to `bbr`, but was concerned of disrupting their current workflow. A method for converting previous model submissions would facilitate the transition. A [prototype PR](https://github.com/metrumresearchgroup/bbr/pull/568) has already been developed, and we have some examples [here](https://github.com/metrumresearchgroup/bbr/blob/67f8c0fe109e84cd2bf5b525f586ef32366518de/inst/dev-examples/convert-psn-to-bbr.R). Given the number of `PsN` users, this could allow us to tap in to a larger audience and potentially bring in many more users, yielding more `bbr` & `bbr.bayes` feedback down the line. 

@seth127 and I talked a few times about this feature, and came to the conclusion that this functionality may not belong within the `bbr` framework. This spurred additional conversations about:

* creating a new package for hosting this functionality, and 
* whether we should expand the package’s intentions to be a bit more general. 

Hence, **`bbr.connectors`**. The idea is to provide methods for _transitioning to_ `bbr`, and potentially additional _connectors_ that interface with other software, such as `mrgsolve` or `stan`. We would start with this `PsN` conversion tool, and plan to expand the package’s functionality if more user requests come in.

### Current scope of proposed new package:

- PsN-to-bbr converter, prototyped in [un-merged bbr PR](https://github.com/metrumresearchgroup/bbr/pull/568)
- Potentially other PsN-to-bbr helpers, for example:
   - maybe something to help move from the PsN Rplots functionality towards MeRGE diagnostics ([R script](https://merge.metrumrg.com/expo/expo1-nonmem-foce/posts/model-diagnostics.html) approach or [parameterized Rmd](https://merge.metrumrg.com/expo/expo1-nonmem-foce/posts/model-diagnostics-param-reports.html) approach)

### Potential future extensions

- `stan` / `cmdstanr` / `Rstan`
   - Mostly thinking of users who have been using these directly and want to move to [`bbr.bayes`](https://github.com/metrumresearchgroup/bbr.bayes)
- `mrgsolve`
   - Maybe… would need to talk to @kylebaron and others about whether this is a need, or even makes sense.
- Other `NONMEM` wrappers
   - Potentially [`Xpose`](https://github.com/UUPharmacometrics/xpose/) and/or @tsahota's [`NMProject`](https://github.com/tsahota/NMproject)? Others?

**Comment by kyleam** [comment]
> Seth and I talked a few times about this feature, and came to the conclusion that this functionality may not belong within the `bbr` framework

It's not obvious to me what the rationale for it not being a good fit for bbr is, so it'd be nice to have that laid out.  (I haven’t looked at the PR too closely, but it doesn’t seem too heavy code-wise.)

> Potential future extensions

I don't have any intuition for what a Stan/cmdstanr/Rstan or mrgsolve connector would look like, but it seems like it'd be a good idea to at least have one of them in hand before deciding that it makes sense to have a connectors package.

**Reply by barrettk** [reply]
>I don't have any intuition for what a Stan/cmdstanr/Rstan or mrgsolve connector would look like, but it seems like it'd be a good idea to at least have one of them in hand before deciding that it makes sense to have a connectors package.

I had a similar thought to this. That maybe we should include _something_ besides `PsN` tooling in its first release. That being said, I think the primary motivation is to provide a landing space for the existing `PsN` convertor, while not boxing us in to exclusively `PsN` tooling. I wouldn't necessarily delay a release just to force another tool to be included, but worth having a conversation about.

**Reply by seth127** [reply]
@kyleam some responses with my current thinking on this are below. To be clear though, this is all very exploratory and at this point we're very open to other perspectives.

> kyleam: It's not obvious to me what the rationale for it not being a good fit for bbr is...

My main thinking here is two-fold:

1) Given the complexity and nuance of everything NONMEM-related, this is likely to see a fair amount of development (both fixes and features) once it starts to get used in the wild. Making a bunch of `bbr` releases with "fixed in bug in `convert_psn()`" or "added support for ____ syntax to `convert_psn()`" would just be messy and not ideal in a number of ways.
2) I think it's likely that this would be used a finite number of times by any given user; as a transition tool, and then they would be on to using `bbr` directly. If I'm right, that would make me think of this as _not_ core `bbr` functionality. That is, it wouldn't be something that any arbitrary `bbr` user (or developer) is likely going to need to know about, and therefore I don't love having it cluttering up the docs/tests/etc.

> kyleam: I don't have any intuition for what a Stan/cmdstanr/Rstan or mrgsolve connector would look like...

For the Stan case(s), I'm thinking about someone who has run some Stan models directly through `cmdstanr` or something else, and would like to be able to continue that analysis in `bbr.bayes`, likely by running `copy_model_from()` on one of these previously run models. Or maybe they switched to using `bbr.bayes` mid-project and just want to easily include those (pre-bbr) models in a run log or something. I'm not sure how much (if any) machinery we would need for that sort of thing, but it crossed my mind and made me think "oh, let's not just call this potential new package `psn2bbr` or something like that."

I'm a lot less clear on the `mrgsolve` use case (and if there even is one), but that's why I tagged @kylebaron to see if he has any vision for that. My guess is that it wouldn't be needed, but as we implement things like [`bbr.bayes::run_sims()`](https://github.com/metrumresearchgroup/bbr.bayes/pull/75) it seems like there might be more situations where `bbr` and `mrgsolve` get more intertwined. (Also, I wanted to get another Kyle tagged in this discussion, and mentioning `mrgsolve` seemed like the shortest path. Just kidding. Sort of.)

**Reply by kyleam** [reply]
> 1. Making a bunch of `bbr` releases with "fixed in bug in `convert_psn()`" or "added support for ____ syntax to `convert_psn()`" would just be messy and not ideal in a number of ways.

I don't think I view that as a problem like you do.  But...

> 2. I think it's likely that this would be used a finite number of times by any given user; as a transition tool [...]

... conceptually I can see this argument.

Anyway, where to lump/split a package is a matter of taste (and depends on how people weigh different tradeoffs).  It's not a strong objection from my end, but I did want to understand better what your thinking was.  Thanks for expanding.

> I'm thinking about someone who has run some Stan models directly through `cmdstanr` or something else, and would like to be able to continue that analysis in `bbr.bayes` [...]

I'm still having trouble wrapping my head around what that would look like. There are so many ways cmdstanr, let alone any other Stan approaches, could be structured.   As I mentioned above, I do think there's value in mapping out some design there before deciding to create a package dedicated to this sort of thing, especially because I'm not sure the cmdstanr idea is worth pursuing.  But of course if you're convinced the PsN functionality should be split into a separate package either way, I see why it doesn't matter so much; if you never add another converter, you mostly just end up with a somewhat confusing name.

**Comment by callistosp** [comment]
additional scope consideration: Pirana/PsN generates PDFs of parameter tables with a single button click. This is probably something we could implement if we leverage pmtables functionality. The parameter table generation in MeRGE is already semi-automated with the param key yaml file.

**Reply by seth127** [reply]
Good call @callistosp . We also have the parameter table package that @KatherineKayMRG @graceannobrien and @michaelmcd18 have been working on. That will hopefully be released in the next month or two, so some integration with that could be a part of this as well.
